myst_nb/glue.py

```diff
@@ -0,0 +1,177 @@
+    outputs = []
+    for cell in ntbk.cells:
+        if cell.cell_type != "code":
+            continue
+
+        # If we have outputs, look for scrapbook metadata and reference the key
+        for output in cell["outputs"]:
+            meta = output.get("metadata", {})
+            if "scrapbook" in meta:
+                this_key = meta["scrapbook"]["name"].replace(mime_prefix, "")
+                if key == this_key:
+                    bundle = output["data"]
+                    bundle = {this_key: val for key, val in bundle.items()}
+                    outputs.append(bundle)
+    if len(outputs) == 0:
+        print(f"Warning: did not find key {this_key} in notebook {path_ntbk}")
+        return
+    if len(outputs) > 1:
+        print(
+            f"Warning: multiple variables found for key: {this_key}. Returning first value."
+        )
+    return outputs[0]
+
+
+def find_all_keys(ntbk, keys=None):
+    """Find all `glue` keys in a notebook and return a dictionary with key: outputs."""
+    if keys is None:
+        keys = {}
+
+    for cell in ntbk.cells:
+        if cell.cell_type != "code":
+            continue
+
+        for output in cell["outputs"]:
+            meta = output.get("metadata", {})
+            if "scrapbook" in meta:
+                this_key = meta["scrapbook"]["name"]
+                if this_key in keys:
+                    SPHINX_LOGGER.warn(
+                        f"Over-writing pre-existing glue key: {this_key}"
+                    )
+                keys[this_key] = output
+    return keys
+
+
+# Role and directive for pasting
+class Paste(SphinxDirective):
+    required_arguments = 1
+    final_argument_whitespace = True
+    has_content = False
+
+    option_spec = {"id": directives.unchanged}
+
+    def run(self):
+        return [PasteNode(self.arguments[0], "directive")]
+
+
+def paste_role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+    return [PasteNode(text, "role")], []
+
+
+# Transform to replace nodes with outputs
+class PasteNode(nodes.container):
+    """Represent a MimeBundle in the Sphinx AST, to be transformed later."""
+
+    def __init__(self, key, kind, rawsource="", *children, **attributes):
+        self.key = key
+        self.kind = kind
+        super().__init__("", **attributes)
+
+
+class PasteNodesToDocutils(SphinxTransform):
+    """Use the builder context to transform a CellOutputNode into Sphinx nodes."""
+
+    default_priority = 700
+
+    def apply(self):
+        glue_data = self.app.env.glue_data
+        for paste_node in self.document.traverse(PasteNode):
+
+            # First check if we have both key:format in the key
+            parts = paste_node.key.rsplit(":", 1)
+            if len(parts) == 2:
+                key, formatting = parts
+            else:
+                key = parts[0]
+                formatting = None
+
+            if key not in glue_data:
+                raise KeyError(
```

Comment:
I think this should use the sphinx logging mechanics, to also report the `paste_node` source file and line number

Reply:
Also, should we be raising exceptions, that block the whole build, or just reporting warnings/errors